server/midddleware/forwarder.go

```diff
@@ -86,9 +120,97 @@ func forwarderUnaryServerInterceptor() grpc.UnaryServerInterceptor {
 		txCtx := api.GetTransaction(ctx, req.(proto.Message))
 
 		if txCtx != nil && txCtx.GetOrigin() != types.MyOrigin {
-			return forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			iface, err := forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			log.Err(err).Str("method", info.FullMethod).Msg("forwarded request")
+			return iface, err
 		}
 
 		return handler(ctx, req)
 	}
 }
+
+func proxyHandler(_ interface{}, serverStream grpc.ServerStream) error {
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		return api.Errorf(api.Code_INTERNAL, "failed to determine method name")
+	}
+
+	outgoingCtx, backendConn, req, resp, err := proxyDirector(serverStream.Context(), fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, &grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, backendConn, fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	s2cErrChan := forwardServerToClient(serverStream, req, clientStream)
+	c2sErrChan := forwardClientToServer(clientStream, resp, serverStream)
+
+	for i := 0; i < 2; i++ {
```

Comment:
is this needed because server chan can return `io.EoF` and we need to wait for `c2sErrChan` to return something?

Reply:
Yes, this read from both channels. It doesn't actually matter which one we are closing first. Refactored, please check again.